src/ui/public/config/config.js

```diff
@@ -129,34 +35,84 @@ module.service('config', function (Private, kbnVersion, kbnIndex, $rootScope, bu
    *                             be stored. Defaults to the config key
    * @return {function} - an unbind function
    */
-  config.$bind = function ($scope, key, property) {
-    if (!property) property = key;
-
-    let update = function () {
-      $scope[property] = config.get(key);
-    };
+  config.$bind = function (scope, key, property = key) {
+    config.on(key, update, scope);
+    function update(value) {
+      scope[property] = value;
+    }
+  };
 
+  config.on = function (key, fn, scope = $rootScope) {
+    const postfix = key === `*` ? `` : `.${key}`;
     update();
-    return _.partial(_.invoke, [
-      $scope.$on('change:config.' + key, update),
-      $scope.$on('init:config', update)
-    ], 'call');
+    scope.$on(`change:config${postfix}`, update);
+    scope.$on(`init:config`, update);
+    function update() {
+      fn(config.get(key));
+    }
   };
 
-  /*****
-   * PRIVATE API
-   *****/
-  function change(key, val, silentAndLocal) {
-    // if the previous updater has already fired, then start over with null
-    if (updater && updater.fired) updater = null;
-    // create a new updater
-    if (!updater) updater = new DelayedUpdater(doc);
-    // return a promise that will be resolved once the action is eventually done
-    return updater.update(key, val, silentAndLocal);
+  function change(key, value) {
+    const oldVal = key in settings ? settings[key].userValue : undefined;
+    const newVal = key in defaults && defaults[key].defaultValue === value ? null : value;
+    const unchanged = oldVal === newVal;
+    if (unchanged) {
+      return Promise.resolve();
+    }
+    const initialVal = config.get(key);
+    localUpdate(key, newVal);
+
+    return delayedUpdate(key, newVal)
+      .then(updatedSettings => {
+        settings = mergeSettings(defaults, updatedSettings);
+      })
+      .catch(reason => {
+        localUpdate(key, initialVal);
+        notify.error(reason);
+      });
   }
 
-  config._vals = function () {
-    return _.cloneDeep(vals);
-  };
+  function localUpdate(key, newVal) {
+    const oldVal = config.get(key);
+    patch(key, newVal);
+    advertise(key, oldVal);
+  }
 
+  function patch(key, value) {
+    if (!(key in settings)) {
+      settings[key] = {};
+    }
+    if (value === null) {
+      delete settings[key].userValue;
+    } else {
+      settings[key].userValue = value;
+    }
+  }
+
+  function advertise(key, oldVal) {
+    const newVal = config.get(key);
+    notify.log(`config change: ${key}: ${oldVal} -> ${newVal}`);
+    $rootScope.$broadcast(`change:config`, settings);
```

Comment:
This `$broadcast` should happen after the more specific one, this way low-level objects listening for any change will react after more specific/targeted ones. Of course this won't always be right, but when I did it locally it fixed an issue where the [index pattern](http://github.com/spalger/kibana/blob/f89defa361b6f5ed1bc12e78c8da122f7b0cb4c8/src/ui/public/index_patterns/_index_pattern.js#L81) was updating before the [field formats registry](http://github.com/spalger/kibana/blob/f89defa361b6f5ed1bc12e78c8da122f7b0cb4c8/src/ui/public/registry/field_formats.js#L14), causing it to get incorrect information.